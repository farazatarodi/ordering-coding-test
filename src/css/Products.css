.products-container {
  width: 80%;
  height: fit-content;
}

.products-content {
  margin: 3rem 0;
  display: flex;
  flex-wrap: wrap;
  width: 100%;
  column-gap: 2%;
  align-self: center;
  border: solid 2px var(--color-neutral);
  box-shadow: var(--shadow);
  border-radius: var(--radius);
}

.products-header {
  width: 100%;
  display: grid;
  grid-template-columns: 10% auto 15% 15% 20%;
  padding: 2rem;
  background-color: var(--color-neutral-light);
  border-radius: var(--radius) var(--radius) 0 0;
}
