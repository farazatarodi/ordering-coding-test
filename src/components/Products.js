// main products page component

// Redux imports
import { useSelector } from 'react-redux';

// component imports
import ProductCard from './ProductCard';
import EmptyCard from './EmptyCard';

// css imports
import '../css/Products.css';
import LoadingCard from './LoadingCard';
import ErrorCard from './ErrorCard';

const Products = () => {
  // get all products from store
  const productsState = useSelector((state) => state.productsReducer);
  const products = productsState.products;
  const loading = productsState.loading;
  const error = productsState.error;

  return (
    <div className="products-container">
      <h1 className="heading heading-1">Products</h1>
      <div className="products-content">
        <div className="products-header heading heading-2">
          <strong>ID</strong>
          <strong>Description</strong>
          <strong>Category</strong>
          <strong>Price</strong>
          <strong>Add to order</strong>
        </div>
        {
          // map products to product components
          loading ? (
            <LoadingCard /> // render loading card if loading
          ) : error ? (
            <ErrorCard error={error} /> // render error card if fetch failed
          ) : products && products.length !== 0 ? (
            products.map((product) => (
              <ProductCard
                key={product.id}
                productId={product.id}
                description={product.description}
                category={product.category}
                price={product.price}
              />
            ))
          ) : (
            <EmptyCard /> // render empty card if no products
          )
        }
      </div>
    </div>
  );
};

export default Products;
