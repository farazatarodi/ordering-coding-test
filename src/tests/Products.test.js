import { render, screen } from '@testing-library/react';

import Products from '../components/Products';

import { Provider } from 'react-redux';

import configureStore from 'redux-mock-store';

test('we see the empty table in products page', () => {
  const initialState = {
    productsReducer: { products: [] },
  };
  const mockStore = configureStore();
  const store = mockStore(initialState);

  render(
    <Provider store={store}>
      <Products />
    </Provider>
  );

  expect(screen.getByText(/id/i)).toBeTruthy();
  expect(screen.queryByText(/â‚¬/i)).toBeNull();
});

test('product card shows all info correctly', () => {
  const initialState = {
    productsReducer: {
      products: [
        {
          id: 'testProductId',
          description: 'testDescription',
          category: 'testCategory',
          price: 'testPrice',
        },
      ],
    },
    optionsReducer: [],
    ordersReducer: { orders: [] },
  };
  const mockStore = configureStore();
  const store = mockStore(initialState);

  render(
    <Provider store={store}>
      <Products />
    </Provider>
  );

  expect(screen.getByText(/testproductid/i)).toBeTruthy();
  expect(screen.getByText(/testdescription/i)).toBeTruthy();
  expect(screen.getByText(/testcategory/i)).toBeTruthy();
  expect(screen.getByText(/testprice/i)).toBeTruthy();
});
