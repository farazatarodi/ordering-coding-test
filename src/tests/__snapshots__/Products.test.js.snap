// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correclty 1`] = `
<div
  className="products-container"
>
  <h1
    className="heading heading-1"
  >
    Products
  </h1>
  <div
    className="products-content"
  >
    <div
      className="products-header heading heading-2"
    >
      <strong>
        ID
      </strong>
      <strong>
        Description
      </strong>
      <strong>
        Category
      </strong>
      <strong>
        Price
      </strong>
      <strong>
        Add to order
      </strong>
    </div>
    <div
      className="card text text-display notification-card"
    >
      Nothing to show
    </div>
  </div>
</div>
`;
